core/src/apps/cardano/sign_tx.py

```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
                     raise wire.ProcessError("Invalid output address!")
 
-                outgoing_coins.append(output.amount)
-                output_addresses.append(output.address)
+            outputs_for_cbor.append([base58.decode(address), amount])
 
-        self.change_addresses = change_addresses
-        self.output_addresses = output_addresses
-        self.outgoing_coins = outgoing_coins
-        self.change_coins = change_coins
-        self.change_derivation_paths = change_derivation_paths
+        return outputs_for_cbor
 
-    def _build_witnesses(self, tx_aux_hash: str):
+    def _build_witnesses(self, tx_aux_hash: bytes) -> List[List[bytes]]:
         witnesses = []
         for input in self.inputs:
-            _, node = derive_address_and_node(self.keychain, input.address_n)
-            message = (
-                b"\x01" + cbor.encode(self.protocol_magic) + b"\x58\x20" + tx_aux_hash
-            )
+            node = self.keychain.derive(input.address_n)
+            message = b"\x58\x20" + tx_aux_hash
+
             signature = ed25519.sign_ext(
                 node.private_key(), node.private_key_ext(), message
             )
-            extended_public_key = (
-                remove_ed25519_prefix(node.public_key()) + node.chain_code()
-            )
-            witnesses.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([extended_public_key, signature])),
-                ]
-            )
-
-        return witnesses
-
-    @staticmethod
-    def compute_fee(input_coins_sum: int, outgoing_coins: list, change_coins: list):
-        outgoing_coins_sum = sum(outgoing_coins)
-        change_coins_sum = sum(change_coins)
-
-        return input_coins_sum - outgoing_coins_sum - change_coins_sum
 
-    def serialise_tx(self):
+            public_key = remove_ed25519_prefix(node.public_key())
 
-        self._process_outputs()
+            # todo: GK - verify this works with Byron inputs (after IOHK confirmation/testnet support)
+            witnesses.append([public_key, signature])
 
-        inputs_cbor = []
-        for input in self.inputs:
-            inputs_cbor.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([input.prev_hash, input.prev_index])),
-                ]
-            )
-
-        inputs_cbor = cbor.IndefiniteLengthArray(inputs_cbor)
-
-        outputs_cbor = []
-        for index, address in enumerate(self.output_addresses):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.outgoing_coins[index]]
-            )
+        return witnesses
 
-        for index, address in enumerate(self.change_addresses):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.change_coins[index]]
-            )
+    def serialise_tx(self) -> Tuple[bytes, bytes]:
```

Comment:
I would split this into part that computes tx_body/tx_hash and the rest which does signing/witnesses

Reply:
I've extracted `_build_tx_body()` and `_hash_tx_body()` but I've kept it in the `serialise_tx()` becuase I didn't like the idea of having to make two calls unnecessarily. Let me know if that's better.

As I was doing this it occured to me, that we maybe don't even need the Transaction class anymore. What do you think? Also @refi93 @janmazak ?